src/tests/ipc_tests.rs

```diff
@@ -0,0 +1,274 @@
+#[cfg(test)]
+pub mod ipc_tests {
+    use crate::interface;
+    use crate::safeposix::{cage::*, filesystem, dispatcher::*};
+    use super::super::*;
+    use std::os::unix::fs::PermissionsExt;
+    use std::fs::OpenOptions;
+    use std::time::Instant;
+
+    //#[test]
+    pub fn test_ipc() {
+        // These can't really run until we figure out a better testing system/fsutils
+        //ut_lind_ipc_pipefile();
+        //ut_lind_ipc_pipe();
+        ut_lind_ipc_domain_socket();
+        ut_lind_ipc_socketpair(); 
+    }
+
+
+    pub fn ut_lind_ipc_pipefile() {
+        let byte_chunk: usize = 131072;
+        let num_writes: usize = 8192;
+
+        lindrustinit(0);
+
+        let cage = interface::cagetable_getref(1);
+
+
+        let filefd = cage.open_syscall("test1gb.txt", O_CREAT | O_WRONLY, S_IRWXA);
+
+        let mut buf: Vec<u8> = vec!['A' as u8; byte_chunk];
+        let bufptr = buf.as_mut_ptr();
+
+        for _i in 0..num_writes {
+            cage.write_syscall(filefd, bufptr, byte_chunk);
+        }
+
+        assert_eq!(cage.close_syscall(filefd), 0);
+        assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        lindrustfinalize();
+    }
+    pub fn ut_lind_ipc_pipe() {
+
+        let byte_chunk: usize = 131072;
+        let num_writes: usize = 8192;
+
+        lindrustinit(0);
+
+        let cage1 = interface::cagetable_getref(1);
+
+        let mut pipefds = PipeArray {readfd: -1, writefd: -1};
+        assert_eq!(cage1.pipe_syscall(&mut pipefds), 0);
+        assert_eq!(cage1.fork_syscall(2), 0);
+
+        let sender = std::thread::spawn(move || {
+
+            let cage2 = interface::cagetable_getref(2);
+
+            assert_eq!(cage2.close_syscall(pipefds.writefd), 0);
+            assert_eq!(cage2.dup2_syscall(pipefds.readfd, 0), 0);
+            assert_eq!(cage2.close_syscall(pipefds.readfd), 0);
+
+
+            let mut bytes_read: usize = 1;
+
+            let mut buf: Vec<u8> = Vec::with_capacity(byte_chunk * num_writes);
+            let mut bufptr = buf.as_mut_ptr();
+            let mut buflen: usize = 0;
+
+            while bytes_read != 0 {
+                bytes_read = cage2.read_syscall(0, bufptr, byte_chunk) as usize;
+                unsafe {
+                    bufptr = bufptr.add(bytes_read);
+                    buf.set_len(buflen + bytes_read);
+                    buflen += bytes_read;
+                }
```

Comment:
In Rust, we don't want to use unsafe code blocks unless we have to. I think we can make this unsafe code block more secure as follows:

```suggestion
                bufptr = bufptr.add(bytes_read);
                buf.resize(buflen + bytes_read, 0);
                buflen += bytes_read;
```

Reply:
The resize can be move out, this makes sense. But the pointer arithmetic operation still have to be in an unsafe block.